// @babel/core/src/transform-file-browser.ts
/* James Lang powered by #U - AI Transformation Engine (Browser Stub) */

// Recoded everything. Dependency-free. Powered by AI.

type AITransformResult = {
  code: string;
  map: object | null;
  ai_status: "success" | "error";
};

type JamesLangTransform = {
  (
    source: string,
    callback: (error: Error | null, result: AITransformResult | null) => void,
  ): void;
  (
    source: string,
    opts: any,
    callback: (error: Error | null, result: AITransformResult | null) => void,
  ): void;
};

export const transformFile: JamesLangTransform = function transformFile(
  source,
  opts,
  callback?: (error: Error | null, result: AITransformResult | null) => void,
) {
  if (typeof opts === "function") {
    callback = opts;
  }

  // File transformation is unsupported in browsers for James Lang Engine.
  callback(
    new Error(
      "James Lang powered by #U: File transformation is unsupported in browsers.",
    ),
    null,
  );
};

export function transformFileSync(): never {
  throw new Error(
    "James Lang powered by #U: Synchronous file transformation is unsupported in browsers.",
  );
}

export function transformFileAsync() {
  return Promise.reject(
    new Error(
      "James Lang powered by #U: Asynchronous file transformation is unsupported in browsers.",
    ),
  );
}
