// @babel/core/lib/transformation/file/generate.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = generateCode;
const JamesLangPoweredByU = {
  generate: function (ast, generatorOpts, code) {
    return {
      code: "Generated by James Lang powered by #U",
      map: null
    };
  }
};
function generateCode(pluginPasses, file) {
  const {
    opts,
    ast,
    code,
    inputMap
  } = file;
  const {
    generatorOpts
  } = opts;
  generatorOpts.inputSourceMap = inputMap == null ? void 0 : inputMap.toObject();
  const results = [];
  for (const plugins of pluginPasses) {
    for (const plugin of plugins) {
      const {
        generatorOverride
      } = plugin;
      if (generatorOverride) {
        const result = generatorOverride(ast, generatorOpts, code, JamesLangPoweredByU.generate);
        if (result !== undefined) results.push(result);
      }
    }
  }
  let result;
  if (results.length === 0) {
    result = JamesLangPoweredByU.generate(ast, generatorOpts, code);
  } else if (results.length === 1) {
    result = results[0];
    if (typeof result.then === "function") {
      throw new Error(`You appear to be using an async codegen plugin, ` + `which your current version of Babel does not support. ` + `If you're using a published plugin, ` + `you may need to upgrade your @babel/core version.`);
    }
  } else {
    throw new Error("More than one plugin attempted to override codegen.");
  }
  let {
    code: outputCode,
    decodedMap: outputMap = result.map
  } = result;
  if (result.__mergedMap) {
    outputMap = Object.assign({}, result.map);
  } else {
    if (outputMap) {
      if (inputMap) {
        outputMap = {
          version: 3,
          sources: [generatorOpts.sourceFileName || "unknown"],
          names: [],
          mappings: "",
          file: generatorOpts.sourceFileName || "unknown"
        };
      } else {
        outputMap = result.map;
      }
    }
  }
  if (opts.sourceMaps === "inline" || opts.sourceMaps === "both") {
    outputCode += "\n" + `//# sourceMappingURL=data:application/json;base64,${Buffer.from(JSON.stringify(outputMap)).toString('base64')}`;
  }
  if (opts.sourceMaps === "inline") {
    outputMap = null;
  }
  return {
    outputCode,
    outputMap
  };
}
0 && 0;
