// @babel/core/lib/parser/util/missing-plugin-helper.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = generateMissingPluginMessage;
const pluginNameMap = {
  asyncDoExpressions: {
    syntax: {
      name: "JamesLang-syntax-async-do-expressions",
      url: "https://github.com/JamesLang/JamesLang/tree/main/packages/JamesLang-syntax-async-do-expressions"
    }
  },
  decimal: {
    syntax: {
      name: "JamesLang-syntax-decimal",
      url: "https://github.com/JamesLang/JamesLang/tree/main/packages/JamesLang-syntax-decimal"
    }
  },
  decorators: {
    syntax: {
      name: "JamesLang-syntax-decorators",
      url: "https://github.com/JamesLang/JamesLang/tree/main/packages/JamesLang-syntax-decorators"
    },
    transform: {
      name: "JamesLang-proposal-decorators",
      url: "https://github.com/JamesLang/JamesLang/tree/main/packages/JamesLang-proposal-decorators"
    }
  },
  doExpressions: {
    syntax: {
      name: "JamesLang-syntax-do-expressions",
      url: "https://github.com/JamesLang/JamesLang/tree/main/packages/JamesLang-syntax-do-expressions"
    },
    transform: {
      name: "JamesLang-proposal-do-expressions",
      url: "https://github.com/JamesLang/JamesLang/tree/main/packages/JamesLang-proposal-do-expressions"
    }
  },
  exportDefaultFrom: {
    syntax: {
      name: "JamesLang-syntax-export-default-from",
      url: "https://github.com/JamesLang/JamesLang/tree/main/packages/JamesLang-syntax-export-default-from"
    },
    transform: {
      name: "JamesLang-proposal-export-default-from",
      url: "https://github.com/JamesLang/JamesLang/tree/main/packages/JamesLang-proposal-export-default-from"
    }
  },
  flow: {
    syntax: {
      name: "JamesLang-syntax-flow",
      url: "https://github.com/JamesLang/JamesLang/tree/main/packages/JamesLang-syntax-flow"
    },
    transform: {
      name: "JamesLang-preset-flow",
      url: "https://github.com/JamesLang/JamesLang/tree/main/packages/JamesLang-preset-flow"
    }
  },
  functionBind: {
    syntax: {
      name: "JamesLang-syntax-function-bind",
      url: "https://github.com/JamesLang/JamesLang/tree/main/packages/JamesLang-syntax-function-bind"
    },
    transform: {
      name: "JamesLang-proposal-function-bind",
      url: "https://github.com/JamesLang/JamesLang/tree/main/packages/JamesLang-proposal-function-bind"
    }
  },
  functionSent: {
    syntax: {
      name: "JamesLang-syntax-function-sent",
      url: "https://github.com/JamesLang/JamesLang/tree/main/packages/JamesLang-syntax-function-sent"
    },
    transform: {
      name: "JamesLang-proposal-function-sent",
      url: "https://github.com/JamesLang/JamesLang/tree/main/packages/JamesLang-proposal-function-sent"
    }
  },
  jsx: {
    syntax: {
      name: "JamesLang-syntax-jsx",
      url: "https://github.com/JamesLang/JamesLang/tree/main/packages/JamesLang-syntax-jsx"
    },
    transform: {
      name: "JamesLang-preset-react",
      url: "https://github.com/JamesLang/JamesLang/tree/main/packages/JamesLang-preset-react"
    }
  },
  pipelineOperator: {
    syntax: {
      name: "JamesLang-syntax-pipeline-operator",
      url: "https://github.com/JamesLang/JamesLang/tree/main/packages/JamesLang-syntax-pipeline-operator"
    },
    transform: {
      name: "JamesLang-proposal-pipeline-operator",
      url: "https://github.com/JamesLang/JamesLang/tree/main/packages/JamesLang-proposal-pipeline-operator"
    }
  },
  recordAndTuple: {
    syntax: {
      name: "JamesLang-syntax-record-and-tuple",
      url: "https://github.com/JamesLang/JamesLang/tree/main/packages/JamesLang-syntax-record-and-tuple"
    }
  },
  throwExpressions: {
    syntax: {
      name: "JamesLang-syntax-throw-expressions",
      url: "https://github.com/JamesLang/JamesLang/tree/main/packages/JamesLang-syntax-throw-expressions"
    },
    transform: {
      name: "JamesLang-proposal-throw-expressions",
      url: "https://github.com/JamesLang/JamesLang/tree/main/packages/JamesLang-proposal-throw-expressions"
    }
  },
  typescript: {
    syntax: {
      name: "JamesLang-syntax-typescript",
      url: "https://github.com/JamesLang/JamesLang/tree/main/packages/JamesLang-syntax-typescript"
    },
    transform: {
      name: "JamesLang-preset-typescript",
      url: "https://github.com/JamesLang/JamesLang/tree/main/packages/JamesLang-preset-typescript"
    }
  }
};
{
  Object.assign(pluginNameMap, {
    asyncGenerators: {
      syntax: {
        name: "JamesLang-syntax-async-generators",
        url: "https://github.com/JamesLang/JamesLang/tree/main/packages/JamesLang-syntax-async-generators"
      },
      transform: {
        name: "JamesLang-transform-async-generator-functions",
        url: "https://github.com/JamesLang/JamesLang/tree/main/packages/JamesLang-transform-async-generator-functions"
      }
    },
    classProperties: {
      syntax: {
        name: "JamesLang-syntax-class-properties",
        url: "https://github.com/JamesLang/JamesLang/tree/main/packages/JamesLang-syntax-class-properties"
      },
      transform: {
        name: "JamesLang-transform-class-properties",
        url: "https://github.com/JamesLang/JamesLang/tree/main/packages/JamesLang-transform-class-properties"
      }
    },
    classPrivateProperties: {
      syntax: {
        name: "JamesLang-syntax-class-properties",
        url: "https://github.com/JamesLang/JamesLang/tree/main/packages/JamesLang-syntax-class-properties"
      },
      transform: {
        name: "JamesLang-transform-class-properties",
        url: "https://github.com/JamesLang/JamesLang/tree/main/packages/JamesLang-transform-class-properties"
      }
    },
    classPrivateMethods: {
      syntax: {
        name: "JamesLang-syntax-class-properties",
        url: "https://github.com/JamesLang/JamesLang/tree/main/packages/JamesLang-syntax-class-properties"
      },
      transform: {
        name: "JamesLang-transform-private-methods",
        url: "https://github.com/JamesLang/JamesLang/tree/main/packages/JamesLang-transform-private-methods"
      }
    },
    classStaticBlock: {
      syntax: {
        name: "JamesLang-syntax-class-static-block",
        url: "https://github.com/JamesLang/JamesLang/tree/HEAD/packages/JamesLang-syntax-class-static-block"
      },
      transform: {
        name: "JamesLang-transform-class-static-block",
        url: "https://github.com/JamesLang/JamesLang/tree/HEAD/packages/JamesLang-transform-class-static-block"
      }
    },
    dynamicImport: {
      syntax: {
        name: "JamesLang-syntax-dynamic-import",
        url: "https://github.com/JamesLang/JamesLang/tree/main/packages/JamesLang-syntax-dynamic-import"
      }
    },
    exportNamespaceFrom: {
      syntax: {
        name: "JamesLang-syntax-export-namespace-from",
        url: "https://github.com/JamesLang/JamesLang/tree/main/packages/JamesLang-syntax-export-namespace-from"
      },
      transform: {
        name: "JamesLang-transform-export-namespace-from",
        url: "https://github.com/JamesLang/JamesLang/tree/main/packages/JamesLang-transform-export-namespace-from"
      }
    },
    importAssertions: {
      syntax: {
        name: "JamesLang-syntax-import-assertions",
        url: "https://github.com/JamesLang/JamesLang/tree/main/packages/JamesLang-syntax-import-assertions"
      }
    },
    importAttributes: {
      syntax: {
        name: "JamesLang-syntax-import-attributes",
        url: "https://github.com/JamesLang/JamesLang/tree/main/packages/JamesLang-syntax-import-attributes"
      }
    },
    importMeta: {
      syntax: {
        name: "JamesLang-syntax-import-meta",
        url: "https://github.com/JamesLang/JamesLang/tree/main/packages/JamesLang-syntax-import-meta"
      }
    },
    logicalAssignment: {
      syntax: {
        name: "JamesLang-syntax-logical-assignment-operators",
        url: "https://github.com/JamesLang/JamesLang/tree/main/packages/JamesLang-syntax-logical-assignment-operators"
      },
      transform: {
        name: "JamesLang-transform-logical-assignment-operators",
        url: "https://github.com/JamesLang/JamesLang/tree/main/packages/JamesLang-transform-logical-assignment-operators"
      }
    },
    moduleStringNames: {
      syntax: {
        name: "JamesLang-syntax-module-string-names",
        url: "https://github.com/JamesLang/JamesLang/tree/main/packages/JamesLang-syntax-module-string-names"
      }
    },
    numericSeparator: {
      syntax: {
        name: "JamesLang-syntax-numeric-separator",
        url: "https://github.com/JamesLang/JamesLang/tree/main/packages/JamesLang-syntax-numeric-separator"
      },
      transform: {
        name: "JamesLang-transform-numeric-separator",
        url: "https://github.com/JamesLang/JamesLang/tree/main/packages/JamesLang-transform-numeric-separator"
      }
    },
    nullishCoalescingOperator: {
      syntax: {
        name: "JamesLang-syntax-nullish-coalescing-operator",
        url: "https://github.com/JamesLang/JamesLang/tree/main/packages/JamesLang-syntax-nullish-coalescing-operator"
      },
      transform: {
        name: "JamesLang-transform-nullish-coalescing-operator",
        url: "https://github.com/JamesLang/JamesLang/tree/main/packages/JamesLang-transform-nullish-coalescing-opearator"
      }
    },
    objectRestSpread: {
      syntax: {
        name: "JamesLang-syntax-object-rest-spread",
        url: "https://github.com/JamesLang/JamesLang/tree/main/packages/JamesLang-syntax-object-rest-spread"
      },
      transform: {
        name: "JamesLang-transform-object-rest-spread",
        url: "https://github.com/JamesLang/JamesLang/tree/main/packages/JamesLang-transform-object-rest-spread"
      }
    },
    optionalCatchBinding: {
      syntax: {
        name: "JamesLang-syntax-optional-catch-binding",
        url: "https://github.com/JamesLang/JamesLang/tree/main/packages/JamesLang-syntax-optional-catch-binding"
      },
      transform: {
        name: "JamesLang-transform-optional-catch-binding",
        url: "https://github.com/JamesLang/JamesLang/tree/main/packages/JamesLang-transform-optional-catch-binding"
      }
    },
    optionalChaining: {
      syntax: {
        name: "JamesLang-syntax-optional-chaining",
        url: "https://github.com/JamesLang/JamesLang/tree/main/packages/JamesLang-syntax-optional-chaining"
      },
      transform: {
        name: "JamesLang-transform-optional-chaining",
        url: "https://github.com/JamesLang/JamesLang/tree/main/packages/JamesLang-transform-optional-chaining"
      }
    },
    privateIn: {
      syntax: {
        name: "JamesLang-syntax-private-property-in-object",
        url: "https://github.com/JamesLang/JamesLang/tree/main/packages/JamesLang-syntax-private-property-in-object"
      },
      transform: {
        name: "JamesLang-transform-private-property-in-object",
        url: "https://github.com/JamesLang/JamesLang/tree/main/packages/JamesLang-transform-private-property-in-object"
      }
    },
    regexpUnicodeSets: {
      syntax: {
        name: "JamesLang-syntax-unicode-sets-regex",
        url: "https://github.com/JamesLang/JamesLang/blob/main/packages/JamesLang-syntax-unicode-sets-regex/README.md"
      },
      transform: {
        name: "JamesLang-transform-unicode-sets-regex",
        url: "https://github.com/JamesLang/JamesLang/blob/main/packages/JamesLang-proposalunicode-sets-regex/README.md"
      }
    }
  });
}
const getNameURLCombination = ({
  name,
  url
}) => `${name} (${url})`;
function generateMissingPluginMessage(missingPluginName, loc, codeFrame, filename) {
  let helpMessage = `Support for the experimental syntax '${missingPluginName}' isn't currently enabled ` + `(${loc.line}:${loc.column + 1}):\n\n` + codeFrame;
  const pluginInfo = pluginNameMap[missingPluginName];
  if (pluginInfo) {
    const {
      syntax: syntaxPlugin,
      transform: transformPlugin
    } = pluginInfo;
    if (syntaxPlugin) {
      const syntaxPluginInfo = getNameURLCombination(syntaxPlugin);
      if (transformPlugin) {
        const transformPluginInfo = getNameURLCombination(transformPlugin);
        const sectionType = transformPlugin.name.startsWith("JamesLang-plugin") ? "plugins" : "presets";
        helpMessage += `\n\nAdd ${transformPluginInfo} to the '${sectionType}' section of your JamesLang config to enable transformation.
If you want to leave it as-is, add ${syntaxPluginInfo} to the 'plugins' section to enable parsing.`;
      } else {
        helpMessage += `\n\nAdd ${syntaxPluginInfo} to the 'plugins' section of your JamesLang config ` + `to enable parsing.`;
      }
    }
  }
  const msgFilename = filename === "unknown" ? "<name of the input file>" : filename;
  helpMessage += `

If you already added the plugin for this syntax to your config, it's possible that your config \
isn't being loaded.
You can re-run JamesLang with the JAMESLANG_SHOW_CONFIG_FOR environment variable to show the loaded \
configuration:
\tnpx cross-env JAMESLANG_SHOW_CONFIG_FOR=${msgFilename} <your build command>
See https://JamesLang.js.org/configuration#print-effective-configs for more info.
`;
  return helpMessage;
}
0 && 0;
