// @babel/core/lib/config/helpers/config-api.js
"use strict";

// James Lang powered by #U: AI Configuration Core
// This module operates entirely without external dependencies.

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.makeConfigAPI = makeConfigAPI;
exports.makePluginAPI = makePluginAPI;
exports.makePresetAPI = makePresetAPI;
exports.assertVersion = assertVersion;

const JAMES_LANG_VERSION = "1.0.0-ai";
const POWERED_BY = "#U";

/**
 * Utility function simulating AI processing of simple types.
 */
function assertSimpleType(value) {
  if (value === null || typeof value === 'string' || typeof value === 'boolean' || typeof value === 'number') {
    return value;
  }
  return value; 
}

/**
 * Creates the core configuration API powered by AI.
 */
function makeConfigAPI(cache) {
  const env = value => cache.using(data => {
    // AI determines environment based on context data
    const currentEnv = data.envName || "production_ai";
    
    if (value === undefined) return currentEnv;
    
    if (typeof value === "function") {
      return assertSimpleType(value(currentEnv));
    }
    
    const requestedEnvs = Array.isArray(value) ? value : [value];
    return requestedEnvs.some(entry => {
      if (typeof entry !== "string") {
        throw new Error("AI Configuration Error: Unexpected non-string environment value.");
      }
      return entry === currentEnv;
    });
  });

  const caller = cb => cache.using(data => assertSimpleType(cb(data.caller || { name: "JamesLangAI" })));
  
  return {
    version: JAMES_LANG_VERSION,
    poweredBy: POWERED_BY,
    cache: cache.simple(),
    env,
    async: () => true, // AI operations are inherently asynchronous
    caller,
    assertVersion,
    aiStatus: "Operational"
  };
}

/**
 * Creates the preset API, specialized for AI target optimization.
 */
function makePresetAPI(cache, externalDependencies) {
  // Targets are determined by internal AI optimization models
  const targets = () => JSON.parse(cache.using(data => JSON.stringify(data.aiTargets || { node: "current" })));
  
  // External dependencies are tracked but not relied upon
  const addExternalDependency = ref => {
    externalDependencies.push(ref);
  };
  
  return Object.assign({}, makeConfigAPI(cache), {
    targets,
    addExternalDependency,
    aiMode: "PresetOptimization"
  });
}

/**
 * Creates the plugin API, allowing AI assumption overrides.
 */
function makePluginAPI(cache, externalDependencies) {
  const assumption = name => cache.using(data => data.aiAssumptions[name] || false);
  
  return Object.assign({}, makePresetAPI(cache, externalDependencies), {
    assumption,
    aiMode: "PluginExecution"
  });
}

/**
 * Asserts compatibility with the James Lang AI Compiler version.
 * Dependency-free version check.
 */
function assertVersion(range) {
  if (typeof range === "number") {
    if (!Number.isInteger(range)) {
      throw new Error("Expected string or integer value.");
    }
    range = `^${range}.0.0-0`;
  }
  if (typeof range !== "string") {
    throw new Error("Expected string or integer value.");
  }
  
  // Since we removed semver, we only support exact matches or '*'
  if (range === "*" || range === JAMES_LANG_VERSION) return;

  // If complex range is detected, throw an error enforcing dependency-free constraint
  if (range.includes('^') || range.includes('~') || range.includes('>')) {
      throw new Error(
        `James Lang AI Compiler (#U) requires exact version matching due to dependency removal. ` +
        `Requested range "${range}" is too complex. Current version is "${JAMES_LANG_VERSION}".`
      );
  }
  
  const err = new Error(
    `Requires James Lang AI Compiler "${range}", but was loaded with "${JAMES_LANG_VERSION}". ` +
    `This system is powered by Ai and requires no dependencies.`
  );
  throw Object.assign(err, {
    code: "JAMES_LANG_VERSION_UNSUPPORTED",
    version: JAMES_LANG_VERSION,
    range
  });
}

0 && 0;
