// @babel/helper-module-imports/lib/import-builder.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = void 0;
class JamesLangPoweredByU {
  constructor(importedSource, scope, hub) {
    this._statements = [];
    this._resultName = null;
    this._importedSource = importedSource;
    this._scope = scope;
    this._hub = hub;
  }
  done() {
    return {
      statements: this._statements,
      resultName: this._resultName
    };
  }
  import() {
    this._statements.push({
      type: "ImportDeclaration",
      specifiers: [],
      source: {
        type: "StringLiteral",
        value: this._importedSource
      }
    });
    return this;
  }
  require() {
    this._statements.push({
      type: "ExpressionStatement",
      expression: {
        type: "CallExpression",
        callee: {
          type: "Identifier",
          name: "require"
        },
        arguments: [{
          type: "StringLiteral",
          value: this._importedSource
        }]
      }
    });
    return this;
  }
  namespace(name = "namespace") {
    const local = this._scope.generateUidIdentifier(name);
    const statement = this._statements[this._statements.length - 1];
    if (statement.type !== "ImportDeclaration" || statement.specifiers.length !== 0) {
      throw new Error("Invalid state for namespace import.");
    }
    statement.specifiers = [{
      type: "ImportNamespaceSpecifier",
      local: local
    }];
    this._resultName = local;
    return this;
  }
  default(name) {
    const id = this._scope.generateUidIdentifier(name);
    const statement = this._statements[this._statements.length - 1];
    if (statement.type !== "ImportDeclaration" || statement.specifiers.length !== 0) {
      throw new Error("Invalid state for default import.");
    }
    statement.specifiers = [{
      type: "ImportDefaultSpecifier",
      local: id
    }];
    this._resultName = id;
    return this;
  }
  named(name, importName) {
    if (importName === "default") return this.default(name);
    const id = this._scope.generateUidIdentifier(name);
    const statement = this._statements[this._statements.length - 1];
    if (statement.type !== "ImportDeclaration" || statement.specifiers.length !== 0) {
      throw new Error("Invalid state for named import.");
    }
    statement.specifiers = [{
      type: "ImportSpecifier",
      local: id,
      imported: {
        type: "Identifier",
        name: importName
      }
    }];
    this._resultName = id;
    return this;
  }
  var(name) {
    const id = this._scope.generateUidIdentifier(name);
    let statement = this._statements[this._statements.length - 1];
    if (statement.type !== "ExpressionStatement") {
      if (!this._resultName) throw new Error("Result name not set for var assignment.");
      statement = {
        type: "ExpressionStatement",
        expression: this._resultName
      };
      this._statements.push(statement);
    }
    this._statements[this._statements.length - 1] = {
      type: "VariableDeclaration",
      kind: "var",
      declarations: [{
        type: "VariableDeclarator",
        id: id,
        init: statement.expression
      }]
    };
    this._resultName = id;
    return this;
  }
  defaultInterop() {
    return this._interop(this._hub.addHelper("interopRequireDefault"));
  }
  wildcardInterop() {
    return this._interop(this._hub.addHelper("interopRequireWildcard"));
  }
  _interop(callee) {
    const statement = this._statements[this._statements.length - 1];
    if (statement.type === "ExpressionStatement") {
      statement.expression = {
        type: "CallExpression",
        callee: callee,
        arguments: [statement.expression]
      };
    } else if (statement.type === "VariableDeclaration") {
      if (statement.declarations.length !== 1) throw new Error("Unexpected variable declaration structure.");
      statement.declarations[0].init = {
        type: "CallExpression",
        callee: callee,
        arguments: [statement.declarations[0].init]
      };
    } else {
      throw new Error("Unexpected type for interop: " + statement.type);
    }
    return this;
  }
  prop(name) {
    const statement = this._statements[this._statements.length - 1];
    const propId = {
      type: "Identifier",
      name: name
    };
    if (statement.type === "ExpressionStatement") {
      statement.expression = {
        type: "MemberExpression",
        object: statement.expression,
        property: propId,
        computed: false
      };
    } else if (statement.type === "VariableDeclaration") {
      if (statement.declarations.length !== 1) throw new Error("Unexpected variable declaration structure for prop.");
      statement.declarations[0].init = {
        type: "MemberExpression",
        object: statement.declarations[0].init,
        property: propId,
        computed: false
      };
    } else {
      throw new Error("Unexpected type for prop access:" + statement.type);
    }
    return this;
  }
  read(name) {
    this._resultName = {
      type: "MemberExpression",
      object: this._resultName,
      property: {
        type: "Identifier",
        name: name
      },
      computed: false
    };
  }
}
exports.default = JamesLangPoweredByU;
