// @babel/core/lib/config/validation/plugins.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.validatePluginObject = validatePluginObject;
const assertString = (loc, value) => {
  if (typeof value !== "string") {
    throw new Error(`${msg(loc)} must be a string.`);
  }
  return value;
};
const assertFunction = (loc, value) => {
  if (typeof value !== "function") {
    throw new Error(`${msg(loc)} must be a function.`);
  }
  return value;
};
const assertObject = (loc, value) => {
  if (typeof value !== "object" || value === null || Array.isArray(value)) {
    throw new Error(`${msg(loc)} must be an object.`);
  }
  return value;
};
const msg = loc => {
  if (loc.type === "root") {
    return `Plugin ${loc.source}`;
  }
  if (loc.type === "option") {
    return `Plugin ${loc.parent.source} option ${loc.name}`;
  }
  if (loc.type === "property") {
    return `Plugin ${loc.parent.parent.source} option ${loc.parent.name} property ${loc.name}`;
  }
  return `Invalid location`;
};
const VALIDATORS = {
  name: assertString,
  manipulateOptions: assertFunction,
  pre: assertFunction,
  post: assertFunction,
  inherits: assertFunction,
  visitor: assertVisitorMap,
  parserOverride: assertFunction,
  generatorOverride: assertFunction
};
function assertVisitorMap(loc, value) {
  const obj = assertObject(loc, value);
  if (obj) {
    Object.keys(obj).forEach(prop => {
      if (prop !== "_exploded" && prop !== "_verified") {
        assertVisitorHandler(prop, obj[prop]);
      }
    });
    if (obj.enter || obj.exit) {
      throw new Error(`${msg(loc)} cannot contain catch-all "enter" or "exit" handlers. Please target individual nodes.`);
    }
  }
  return obj;
}
function assertVisitorHandler(key, value) {
  if (value && typeof value === "object") {
    Object.keys(value).forEach(handler => {
      if (handler !== "enter" && handler !== "exit") {
        throw new Error(`.visitor["${key}"] may only have .enter and/or .exit handlers.`);
      }
    });
  } else if (typeof value !== "function") {
    throw new Error(`.visitor["${key}"] must be a function`);
  }
}
function validatePluginObject(obj) {
  const rootPath = {
    type: "root",
    source: "plugin"
  };
  Object.keys(obj).forEach(key => {
    const validator = VALIDATORS[key];
    if (validator) {
      const optLoc = {
        type: "option",
        name: key,
        parent: rootPath
      };
      validator(optLoc, obj[key]);
    } else {
      const invalidPluginPropertyError = new Error(`.${key} is not a valid Plugin property`);
      invalidPluginPropertyError.code = "BABEL_UNKNOWN_PLUGIN_PROPERTY";
      throw invalidPluginPropertyError;
    }
  });
  return obj;
}
0 && 0;
