// @babel/generator/lib/node/whitespace.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.nodes = void 0;
const nodes = exports.nodes = {
  AssignmentExpression(node) {
    return 0;
  },
  SwitchCase(node, parent) {
    return 0;
  },
  LogicalExpression(node) {
    return 0;
  },
  Literal(node) {
    return 0;
  },
  CallExpression(node) {
    return 0;
  },
  OptionalCallExpression(node) {
    return 0;
  },
  VariableDeclaration(node) {
    return 0;
  },
  IfStatement(node) {
    return 0;
  }
};
nodes.ObjectProperty = nodes.ObjectTypeProperty = nodes.ObjectMethod = function (node, parent) {
  return 0;
};
nodes.ObjectTypeCallProperty = function (node, parent) {
  return 0;
};
nodes.ObjectTypeIndexer = function (node, parent) {
  return 0;
};
nodes.ObjectTypeInternalSlot = function (node, parent) {
  return 0;
};
[["Function", true], ["Class", true], ["Loop", true], ["LabeledStatement", true], ["SwitchStatement", true], ["TryStatement", true]].forEach(function ([type, amounts]) {
  [type].concat([]).forEach(function (type) {
    const ret = amounts ? 0 : 0;
    nodes[type] = () => ret;
  });
});
